Modules/Loadable/Markups/Widgets/qSlicerMarkupsAdditionalWidget.h

```diff
@@ -0,0 +1,51 @@
+/*==============================================================================
+
+  Copyright (c) The Intervention Centre
+  Oslo University Hospital, Oslo, Norway. All Rights Reserved.
+
+  See COPYRIGHT.txt
+  or http://www.slicer.org/copyright/copyright.txt for details.
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  This file was originally developed by Rafael Palomar (The Intervention Centre,
+  Oslo University Hospital) and was supported by The Research Council of Norway
+  through the ALive project (grant nr. 311393).
+
+==============================================================================*/
+
+#ifndef __qslicermarkupsadditionalwidget_h_
+#define __qslicermarkupsadditionalwidget_h_
+
+// Qt Slicer includes
+#include "qSlicerWidget.h"
+
+// Markups Widgets includes
+#include "qSlicerMarkupsModuleWidgetsExport.h"
+
+class vtkMRMLMarkupsNode;
+
+class Q_SLICER_MODULE_MARKUPS_WIDGETS_EXPORT
```

Comment:
I don't understand the role of this class.
I would expect that there would be a MarkupsAdditionalOptionsWidget base class that the markups module widget would show whenever a suitable node type is selected.
These "additional options" widgets could be registered along with the markups node and widgetrepresentation, but probably it would be more flexible if they would be registered independently (because then each widget could return in its canManage() method if it is applicable to the selected node and the markups module widget would show/hide accordingly; or the widget could even show/hide/disable itself depending on what node is selected).

Reply:
@lassoan, this class plays the role you describe. `qSlicerMarkupsAdditionalWidget` is an abstract base class for `qSlicerMarkupsCurveSettingsWidget`, `qSlicerMarkupsAngleMeasurementsWidget`, `qMRMLMarkupsROIWidget` and potentially any markups-specific new widgets. These widgets only appear when the corresponding markup is selected.

Note that these additional widgets are not registered using `vtkSlicerMarkupsLogic` but `qSlicerMarkupsWidget`:
```cpp
qSlicerAbstractModuleRepresentation* qSlicerMarkupsModule::createWidgetRepresentation()
{
  // Create and configure module widget.
  qSlicerMarkupsModuleWidget* moduleWidget = new qSlicerMarkupsModuleWidget();

  // Register markups
  moduleWidget->setMarkup("Fiducial");
  moduleWidget->setMarkup("Line");
  moduleWidget->setMarkup("Angle", new qSlicerMarkupsAngleMeasurementsWidget);
  moduleWidget->setMarkup("Curve", new qSlicerMarkupsCurveSettingsWidget);
  moduleWidget->setMarkup("ClosedCurve");
  moduleWidget->setMarkup("Plane");
  moduleWidget->setMarkup("ROI", new qMRMLMarkupsROIWidget);

  // Set the number of columns for the grid of "add markups buttons" to the number of markups
  // regitered in this module.
  moduleWidget->setCreateMarkupsButtonsColumns(7);

  return moduleWidget;
}
```

Reply:
OK. To make things more clear, add documentation to this class, rename `setMarkup` to something that reflects what it actually does, and decouple the markup registration from the additional widget registration, because additional modules may want to register more widgets (without registering new markup types) and a widget may handle multiple markup types (e.g., new import/export or visualization options that can work with several markup types).